### DISCUSSION TITLE
How to use multiple profiles?

### DISCUSSION THREAD

**Discussion by Daniel-Khodabakhsh** [slick/slick#2326] [Questions] [answered]
I have a use-case where I'm using `scala-guice` to inject various services based on some startup variables. I need to be able to either connect to an H2 or PostgreSQL database. According to the documentation, I need to import the database specific profile (`import slick.jdbc.H2Profike.api._`) but this does not suit my use-case since it's not possible to dynamically import at runtime.

I also would like to avoid a code-generation answer as well. Ideally, based on a config or environment variable, I want to be able to boot into either H2 or Postgres.

I've tried injecting a `JdbcProfile#API` with this interface and concrete class:

```scala
import slick.jdbc.JdbcProfile

trait ProfileProvider {
  val api: JdbcProfile#API
}
```

```scala
import slick.jdbc.{H2Profile, JdbcProfile}

class H2ProfileProvider extends ProfileProvider {
  override val api: JdbcProfile#API = H2Profile.api
}
```

and injecting it like so:

```scala
class AccountTable @Inject() (profileProvider: ProfileProvider) {
  import profileProvider.api._

  // Following line has an error
  class AccountTable(tag: Tag) 
      extends Table[AccountEntity](tag, Some("schemaName"), "ACCOUNT") {
    ...
  }
}
```

But I get an error for the line marked above, stating:

> slick.jdbc.JdbcProfile is not a legal prefix for a constructor
>       extends Table[AccountEntity](tag, Some("schemaName"), "ACCOUNT") {

How can I dynamically inject a profile at startup-time?

**Comment by nafg** [comment]
Does it work if you change `def api` to `val api`?

**Reply by Daniel-Khodabakhsh** [reply]
Ah sorry that's a typo, `def` has more issues yes, the error I'm getting is with `val`. I'll update my question

**Comment by nafg** [comment]
I think the issue is you're trying to abstract over the `API` instead of abstracting over the `JdbcProfile` and importing its `.api._`.

You should be able to use a trait like you are but in other regards try to follow the example at https://scala-slick.org/doc/3.3.3/database.html#a-dao-class

Something like this:

```scala
import slick.jdbc.JdbcProfile

trait ProfileProvider {
  val profile: JdbcProfile
}
```

```scala
import slick.jdbc.{H2Profile, JdbcProfile}

class H2ProfileProvider extends ProfileProvider {
  override val profile = H2Profile
}
```

```scala
class AccountTable @Inject() (profileProvider: ProfileProvider) {
  import profileProvider.profile.api._

  // Following line has an error
  class AccountTable(tag: Tag) 
      extends Table[AccountEntity](tag, Some("schemaName"), "ACCOUNT") {
    ...
  }
}

```

**Comment by Daniel-Khodabakhsh** [comment, marked as answer]
With @nafg's answer we do get the ability to dynamically inject the chosen profile, but I had additional augmentations to the `JdbcProfile.api._` which are not shared between the different profiles; this augmentation being array support.

Did some deeper digging into the difference between the two profiles I was using (H2 and Postgres via `slick-pg`) and found that the profile differences, between each other and between `JdbcProfile`, were minimal and not used in our typical queries.

So instead what I ended up doing is just import `JdbcProfile.api._` directly in all of my table files by first creating an object instance:

```scala
package my.project.data.support

import slick.jdbc.JdbcProfile

object JdbcProfile extends JdbcProfile
```

which allows me to import `JdbcProfile.api._` like so:

```scala
import my.project.data.support.JdbcProfile.api._ // Generic import

class AccountTable @Inject() () {
  // No more errors
  class AccountTable(tag: Tag) 
      extends Table[AccountEntity](tag, Some("schemaName"), "ACCOUNT") {
    ...
  }
}
```

To support the augmentations which differed between profiles, I created a `DatabaseSupport` interface:

```scala
package my.project.data.support

/**
  * Database support implicits and methods.
  */
trait DatabaseSupport {

  /**
    * Implicit database array type.
    */
  implicit val seqStringType: JdbcProfile.DriverJdbcType[Seq[String]]

  ...
}
```

where the H2 implementation has a custom implementation of `seqStringType`, and where the Postgres implementation maps to the `slick-pg`'s library implementation.

```scala
implicit override val seqStringType: JdbcProfile.DriverJdbcType[Seq[String]] =
    new pgArrayJdbcTypes.SimpleArrayJdbcType[String]("text").asInstanceOf[JdbcProfile.DriverJdbcType[Seq[String]]]
```

This is then injected when I need to use the a SQL array:

```scala
import my.project.data.support.DatabaseSupport
import my.project.data.support.JdbcProfile.api._

class CleaningDataTable @Inject() (databaseSupport: DatabaseSupport) {
  import databaseSupport._

  class CleaningDataTable(tag: Tag)
    extends Table[CleaningDataEntity](tag, Some("cleaning"), "DATA") {
    ...
    def data: Rep[Seq[String]] = column[Seq[String]]("data") // Can use array type.
    ...
  }
}
```

In these snippets I only outline the SQL array type support, but I ended up needing to support another difference between H2 and Postgres which I also overcame by leveraging the `DatabaseSupport` interface.

Just wanted to share my thoughts on the slick concept of profile: I think it's flawed in that slick should really just offer up generic interfaces that can be used by all table and query definitions, and then use the driver definition to load database-specific implementations (like what MS's EntityFramework appears to do). If there were additional database feature requirements, they could be implemented via a user-defined interface like what I did here. No need to juggle which profile object to use.
